File: librarymanager/librarymanager-be/src/main/java/ru/spring5/elena/librarymanager/services/AuthorService.java
package ru.spring5.elena.librarymanager.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import ru.spring5.elena.librarymanager.api.dto.Author;
import ru.spring5.elena.librarymanager.api.dto.Edition;
import ru.spring5.elena.librarymanager.domain.entity.AuthorEntity;
import ru.spring5.elena.librarymanager.domain.repository.AuthorRepository;
import ru.spring5.elena.librarymanager.domain.repository.EditionRepository;
import ru.spring5.elena.librarymanager.mapper.AuthorMapper;
import ru.spring5.elena.librarymanager.mapper.EditionMapper;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@RequiredArgsConstructor
@Service
@Slf4j

public class AuthorService {

    private final AuthorRepository authorRepository;
    private final EditionRepository editionRepository;

    @Transactional(readOnly = true)
    public List<Author> findAll() {
        return authorRepository.findAll()
                .stream()
                .map(AuthorMapper::map)
                .collect(Collectors.toList());
    }

    @Transactional(readOnly = true)
    public Page<Author> findAllByPage(Pageable pageable) {
        return authorRepository.findAll(pageable)
                .map(AuthorMapper::map);
    }

    @Transactional(readOnly = true)
    public Author findAuthorById(Long authorId) {
        return authorRepository.findById(authorId)
                .map(AuthorMapper::map)
                .orElse(null);
    }
    
    @Transactional
    public Author save(Long authorId,  Author dto) {
    	return authorRepository.findById(authorId)
                .map(entity -> AuthorMapper.map(dto, entity))
                .map(authorRepository::saveAndFlush)
                .map(AuthorMapper::map)
                .get();

    }

    @Transactional
    public Author create(Author dto) {
        return   Optional.of(dto)
                .map(AuthorMapper::map)
                .map(authorRepository::saveAndFlush)
                .map(AuthorMapper::map)
                .get();

    }

    @Transactional(readOnly = true)
    public Page<Edition> findEditionsByAuthorId(Long authorId, Pageable pageable) {
        return authorRepository.findById(authorId)
        	   .map(entity -> editionRepository.findAllByAuthor(entity, pageable))
               .orElse(Page.empty())
        	   .map(EditionMapper::map);
    }
    
    @Transactional
	public List<Edition> updateEditionsByAuthorId(Long authorId, List<Long> editionIds) {
		return authorRepository
				.findById(authorId)
				.map(authorEntity-> {
					authorEntity.getEditions().clear();
					authorEntity.getEditions().addAll(
							editionIds.stream()
  							.map(editionRepository::getOne)
  							.collect(Collectors.toList())
  							);
					return authorEntity;
				 })
				.map(authorRepository::saveAndFlush)
				.map(AuthorEntity :: getEditions)
				.orElse(Collections.emptySet())
				.stream()
				.map(EditionMapper :: map)
				.collect(Collectors.toList());
		
	}

}

File: librarymanager-be/src/main/java/ru/elenacherev/librarymanager/services/AuthorService.kt
package ru.elenacherev.librarymanager.services

import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import ru.elenacherev.librarymanager.api.dto.Author
import ru.elenacherev.librarymanager.api.dto.Edition
import ru.elenacherev.librarymanager.domain.entity.AuthorEntity
import ru.elenacherev.librarymanager.domain.entity.EditionEntity
import ru.elenacherev.librarymanager.domain.repository.AuthorRepository
import ru.elenacherev.librarymanager.domain.repository.EditionRepository
import ru.elenacherev.librarymanager.mapper.map
import java.util.*

@Service
class AuthorService(
    val authorRepository: AuthorRepository,
    val editionRepository: EditionRepository
) {

    @Transactional(readOnly = true)
    fun findAll(): List<Author> {
        return authorRepository
            .findAll()
            .map(AuthorEntity::map)
    }

    @Transactional(readOnly = true)
    fun findAllByPage(pageable: Pageable): Page<Author> {
        return authorRepository
            .findAll(pageable)
            .map(AuthorEntity::map)

    }

    @Transactional(readOnly = true)
    fun findAuthorById(authorId: UUID): Author? {
        return authorRepository.findById(authorId)
            .map(AuthorEntity::map)
            .orElse(null)
    }

    @Transactional
    fun save(authorId: UUID, dto: Author): Author {
        return authorRepository
            .findById(authorId)
            .map(dto::map)
            .map { author: AuthorEntity -> authorRepository.saveAndFlush(author) }
            .map(AuthorEntity::map)
            .get()
    }

    @Transactional
    fun create(dto: Author): Author {
        return Optional.of(dto)
            .map(Author::map)
            .map { author: AuthorEntity -> authorRepository.saveAndFlush(author) }
            .map(AuthorEntity::map)
            .get()
    }

    @Transactional(readOnly = true)
    fun findEditionsByAuthorId(authorId: UUID, pageable: Pageable): Page<Edition> {
        return authorRepository.findById(authorId)
            .map { entity -> editionRepository.findAllByAuthor(entity, pageable) }
            .orElse(Page.empty())
            .map(EditionEntity::map)
    }

    @Transactional
    fun updateEditionsByAuthorId(authorId: UUID, editionIds: List<UUID>): List<Edition> {
        return authorRepository
            .findById(authorId)
            .map { authorEntity ->
                authorEntity.editions.clear()
                authorEntity.editions += editionIds.map(editionRepository::getOne)
                return@map authorEntity
            }
            .map { author: AuthorEntity -> authorRepository.saveAndFlush(author) }
            .map { obj: AuthorEntity -> obj.editions }
            .orElse(mutableSetOf())
            .map(EditionEntity::map)
            .toList()
    }
}